contributors/design-proposals/api-machinery/customresource-conversion-webhook.md

```diff
@@ -0,0 +1,861 @@
+
+    4. CRD authors will need to write a Webhook Host that accepts any version and returns any version.
+
+    5. Toolkits like kube-builder and operator-sdk are expected to provide flows to assist users to generate Webhook Hosts.
+
+
+# Detailed Design
+
+
+## CRD API Changes
+
+The CustomResourceDefinitionSpec is extended to have a new section where webhooks are defined: 
+
+```golang
+// CustomResourceDefinitionSpec describes how a user wants their resource to appear
+type CustomResourceDefinitionSpec struct {
+  Group string
+  Version string
+  Names CustomResourceDefinitionNames
+  Scope ResourceScope
+  // This optional and correspond to the first version in the versions list
+  Validation *CustomResourceValidation
+  // Optional, correspond to the first version in the versions list 
+  Subresources *CustomResourceSubresources
+  Versions []CustomResourceDefinitionVersion
+  // Optional, and correspond to the first version in the versions list  
+  AdditionalPrinterColumns []CustomResourceColumnDefinition
+
+  Conversion *CustomResourceConversion
+}
+
+type CustomResourceDefinitionVersion struct {
+  Name string
+  Served Boolean
+  Storage Boolean
+  // These three fields should not be set for first item in Versions list
+  Schema *JSONSchemaProp
+  Subresources *CustomResourceSubresources
+  AdditionalPrinterColumns []CustomResourceColumnDefinition
+}
+
+Type CustomResourceConversion struct {
+  // Conversion strategy, either "nop” or "webhook”. If webhook is set, Webhook field is required.
+  Strategy string
+
+  // Additional information for external conversion if strategy is set to external
+  // +optional
+  Webhook *CustomResourceConversionWebhook
+}
+
+type CustomResourceConversionWebhook {
+  // ClientConfig defines how to communicate with the webhook. This is the same config used for validating/mutating webhooks.
+  ClientConfig WebhookClientConfig
+}
+```
+
+## Defaulting
+
+In case that there is no versions list, a single version with values defaulted to top level version will be created. That means a single version with a name set to spec.version.
+
+
+## Validation
+
+To keep backward compatibility, the top level fields (schema, additionalPrinterColumns or subresources) are stay the same and source of truth for first (top) version. The first item in the versions list should not set any of those fields. The plan is to use unified version list for v1.
+
+
+## Support Level
+
+The feature will be alpha in the first supported version and will have a feature gate that is defaulted to false. The roll-back story with a feature gate is much more clear. if we have the features as alpha in version Y (>X where the feature is missing) and we make it beta in version Z, it is not safe to use the feature and downgrade from Y to X but the feature is alpha in Y which is fine. It is safe to downgrade from Z to Y (given that we enable the feature gate in Y) and that is desirable as the feature is beta in Z. I will add this detail to the doc.
```

Comment:
s/in the first supported version/in the first implementation/

Reply:
better write "kubernetes release" instead of version. Here are so many versions in the doc with different meaning.

Reply:
s/I will add this detail to the doc./?/

Reply:
done (last part was the copy/paste monkey's mistake :) )